File: anyflix-backend/lib/models/anilist.py
"""AniList API models for GraphQL responses."""

from __future__ import annotations

from enum import Enum
from typing import Any

from pydantic import BaseModel, field_validator


class MediaType(str, Enum):
    """Media type enumeration."""

    ANIME = "ANIME"
    MANGA = "MANGA"


class MediaFormat(str, Enum):
    """Media format enumeration."""

    TV = "TV"
    TV_SHORT = "TV_SHORT"
    MOVIE = "MOVIE"
    SPECIAL = "SPECIAL"
    OVA = "OVA"
    ONA = "ONA"
    MUSIC = "MUSIC"
    MANGA = "MANGA"
    NOVEL = "NOVEL"
    ONE_SHOT = "ONE_SHOT"


class MediaStatus(str, Enum):
    """Media status enumeration."""

    FINISHED = "FINISHED"
    RELEASING = "RELEASING"
    NOT_YET_RELEASED = "NOT_YET_RELEASED"
    CANCELLED = "CANCELLED"
    HIATUS = "HIATUS"


class MediaSeason(str, Enum):
    """Media season enumeration."""

    WINTER = "WINTER"
    SPRING = "SPRING"
    SUMMER = "SUMMER"
    FALL = "FALL"


class MediaSource(str, Enum):
    """Media source enumeration."""

    ORIGINAL = "ORIGINAL"
    MANGA = "MANGA"
    LIGHT_NOVEL = "LIGHT_NOVEL"
    VISUAL_NOVEL = "VISUAL_NOVEL"
    VIDEO_GAME = "VIDEO_GAME"
    OTHER = "OTHER"
    NOVEL = "NOVEL"
    DOUJINSHI = "DOUJINSHI"
    ANIME = "ANIME"
    WEB_NOVEL = "WEB_NOVEL"
    LIVE_ACTION = "LIVE_ACTION"
    GAME = "GAME"
    COMIC = "COMIC"
    MULTIMEDIA_PROJECT = "MULTIMEDIA_PROJECT"
    PICTURE_BOOK = "PICTURE_BOOK"


class RelationType(str, Enum):
    """Relation type enumeration."""

    ADAPTATION = "ADAPTATION"
    PREQUEL = "PREQUEL"
    SEQUEL = "SEQUEL"
    PARENT = "PARENT"
    SIDE_STORY = "SIDE_STORY"
    CHARACTER = "CHARACTER"
    SUMMARY = "SUMMARY"
    ALTERNATIVE = "ALTERNATIVE"
    SPIN_OFF = "SPIN_OFF"
    OTHER = "OTHER"
    SOURCE = "SOURCE"
    COMPILATION = "COMPILATION"
    CONTAINS = "CONTAINS"


class CharacterRole(str, Enum):
    """Character role enumeration."""

    MAIN = "MAIN"
    SUPPORTING = "SUPPORTING"
    BACKGROUND = "BACKGROUND"


class StaffLanguage(str, Enum):
    """Staff language enumeration."""

    JAPANESE = "JAPANESE"
    ENGLISH = "ENGLISH"
    KOREAN = "KOREAN"
    ITALIAN = "ITALIAN"
    SPANISH = "SPANISH"
    PORTUGUESE = "PORTUGUESE"
    FRENCH = "FRENCH"
    GERMAN = "GERMAN"
    HEBREW = "HEBREW"
    HUNGARIAN = "HUNGARIAN"


class ExternalLinkType(str, Enum):
    """External link type enumeration."""

    INFO = "INFO"
    STREAMING = "STREAMING"
    SOCIAL = "SOCIAL"


class RankingType(str, Enum):
    """Ranking type enumeration."""

    RATED = "RATED"
    POPULAR = "POPULAR"


class MediaListStatus(str, Enum):
    """Media list status enumeration."""

    CURRENT = "CURRENT"
    PLANNING = "PLANNING"
    COMPLETED = "COMPLETED"
    DROPPED = "DROPPED"
    PAUSED = "PAUSED"
    REPEATING = "REPEATING"


# Base models
class AniListDate(BaseModel):
    """AniList date model."""

    year: int | None = None
    month: int | None = None
    day: int | None = None


class MediaTitle(BaseModel):
    """Media title model."""

    romaji: str | None = None
    english: str | None = None
    native: str | None = None
    userPreferred: str | None = None


class CoverImage(BaseModel):
    """Cover image model."""

    extraLarge: str | None = None
    large: str | None = None
    medium: str | None = None
    color: str | None = None


class Trailer(BaseModel):
    """Trailer model."""

    id: str | None = None
    site: str | None = None
    thumbnail: str | None = None


class NextAiringEpisode(BaseModel):
    """Next airing episode model."""

    airingAt: int | None = None
    timeUntilAiring: int | None = None
    episode: int | None = None


class MediaTag(BaseModel):
    """Media tag model."""

    id: int
    name: str
    description: str | None = None
    category: str | None = None
    rank: int | None = None
    isGeneralSpoiler: bool | None = None
    isMediaSpoiler: bool | None = None
    isAdult: bool | None = None
    userId: int | None = None


class MediaRanking(BaseModel):
    """Media ranking model."""

    id: int
    rank: int
    type: RankingType
    format: MediaFormat
    year: int | None = None
    season: MediaSeason | None = None
    allTime: bool | None = None
    context: str


class StatusDistribution(BaseModel):
    """Status distribution model."""

    status: MediaListStatus | None = None
    amount: int | None = None


class ScoreDistribution(BaseModel):
    """Score distribution model."""

    score: int | None = None
    amount: int | None = None


class MediaStats(BaseModel):
    """Media statistics model."""

    statusDistribution: list[StatusDistribution] | None = None
    scoreDistribution: list[ScoreDistribution] | None = None


class MediaListEntry(BaseModel):
    """Media list entry model."""

    id: int
    status: MediaListStatus | None = None
    score: float | None = None
    progress: int | None = None
    progressVolumes: int | None = None
    repeat: int | None = None
    priority: int | None = None
    private: bool | None = None
    notes: str | None = None
    hiddenFromStatusLists: bool | None = None
    customLists: dict[str, Any] | None = None
    advancedScores: dict[str, Any] | None = None
    startedAt: AniListDate | None = None
    completedAt: AniListDate | None = None
    updatedAt: int | None = None
    createdAt: int | None = None


class StreamingEpisode(BaseModel):
    """Streaming episode model."""

    title: str | None = None
    thumbnail: str | None = None
    url: str | None = None
    site: str | None = None


class ExternalLink(BaseModel):
    """External link model."""

    id: int
    url: str | None = None
    site: str
    siteId: int | None = None
    type: ExternalLinkType | None = None
    language: str | None = None
    color: str | None = None
    icon: str | None = None
    notes: str | None = None
    isDisabled: bool | None = None


# Character and Staff models
class CharacterName(BaseModel):
    """Character name model."""

    first: str | None = None
    middle: str | None = None
    last: str | None = None
    full: str | None = None
    native: str | None = None
    alternative: list[str] | None = None
    alternativeSpoiler: list[str] | None = None
    userPreferred: str | None = None


class CharacterImage(BaseModel):
    """Character image model."""

    large: str | None = None
    medium: str | None = None


class Character(BaseModel):
    """Character model."""

    id: int
    name: CharacterName | None = None
    image: CharacterImage | None = None
    description: str | None = None
    gender: str | None = None
    dateOfBirth: AniListDate | None = None
    age: str | None = None
    bloodType: str | None = None
    isFavourite: bool | None = None
    isFavouriteBlocked: bool | None = None
    siteUrl: str | None = None


class StaffName(BaseModel):
    """Staff name model."""

    first: str | None = None
    middle: str | None = None
    last: str | None = None
    full: str | None = None
    native: str | None = None
    alternative: list[str] | None = None
    userPreferred: str | None = None


class StaffImage(BaseModel):
    """Staff image model."""

    large: str | None = None
    medium: str | None = None


class Staff(BaseModel):
    """Staff model."""

    id: int
    name: StaffName | None = None
    languageV2: str | None = None
    image: StaffImage | None = None
    description: str | None = None
    primaryOccupations: list[str] | None = None
    gender: str | None = None
    dateOfBirth: AniListDate | None = None
    dateOfDeath: AniListDate | None = None
    age: int | None = None
    yearsActive: list[int] | None = None
    homeTown: str | None = None
    bloodType: str | None = None
    isFavourite: bool | None = None
    isFavouriteBlocked: bool | None = None
    siteUrl: str | None = None


class VoiceActor(BaseModel):
    """Voice actor model."""

    id: int
    name: StaffName | None = None
    language: str | None = None
    languageV2: str | None = None
    image: StaffImage | None = None
    description: str | None = None
    primaryOccupations: list[str] | None = None
    gender: str | None = None
    dateOfBirth: AniListDate | None = None
    dateOfDeath: AniListDate | None = None
    age: int | None = None
    yearsActive: list[int] | None = None
    homeTown: str | None = None
    bloodType: str | None = None
    isFavourite: bool | None = None
    isFavouriteBlocked: bool | None = None
    siteUrl: str | None = None


class CharacterEdge(BaseModel):
    """Character edge model."""

    id: int | None = None
    role: CharacterRole | None = None
    name: str | None = None
    voiceActors: list[VoiceActor] | None = None
    node: Character | None = None


class StaffEdge(BaseModel):
    """Staff edge model."""

    id: int | None = None
    role: str | None = None
    node: Staff | None = None


class CharacterConnection(BaseModel):
    """Character connection model."""

    edges: list[CharacterEdge] | None = None
    nodes: list[Character] | None = None


class StaffConnection(BaseModel):
    """Staff connection model."""

    edges: list[StaffEdge] | None = None
    nodes: list[Staff] | None = None


# Studio models
class Studio(BaseModel):
    """Studio model."""

    id: int
    name: str
    isAnimationStudio: bool | None = None
    siteUrl: str | None = None
    isFavourite: bool | None = None


class StudioEdge(BaseModel):
    """Studio edge model."""

    id: int | None = None
    isMain: bool
    node: Studio | None = None


class StudioConnection(BaseModel):
    """Studio connection model."""

    edges: list[StudioEdge] | None = None
    nodes: list[Studio] | None = None


# User and Review models
class UserAvatar(BaseModel):
    """User avatar model."""

    large: str | None = None
    medium: str | None = None


class User(BaseModel):
    """User model."""

    id: int
    name: str
    avatar: UserAvatar | None = None
    bannerImage: str | None = None
    about: str | None = None
    isFollowing: bool | None = None
    isFollower: bool | None = None
    isBlocked: bool | None = None
    bans: list[Any] | None = None
    options: dict[str, Any] | None = None
    mediaListOptions: dict[str, Any] | None = None
    favourites: dict[str, Any] | None = None
    statistics: dict[str, Any] | None = None
    unreadNotificationCount: int | None = None
    siteUrl: str | None = None
    donatorTier: int | None = None
    donatorBadge: str | None = None
    moderatorRoles: list[str] | None = None
    createdAt: int | None = None
    updatedAt: int | None = None


class Review(BaseModel):
    """Review model."""

    id: int
    userId: int | None = None
    mediaId: int | None = None
    mediaType: MediaType | None = None
    summary: str | None = None
    body: str | None = None
    rating: int | None = None
    ratingAmount: int | None = None
    userRating: int | None = None
    score: int | None = None
    private: bool | None = None
    siteUrl: str | None = None
    createdAt: int | None = None
    updatedAt: int | None = None
    user: User | None = None


class PageInfo(BaseModel):
    """Page info model."""

    total: int | None = None
    perPage: int | None = None
    currentPage: int | None = None
    lastPage: int | None = None
    hasNextPage: bool | None = None


class ReviewConnection(BaseModel):
    """Review connection model."""

    edges: list[Any] | None = None
    nodes: list[Review] | None = None
    pageInfo: PageInfo | None = None


# Recommendation models
class MediaRecommendation(BaseModel):
    """Media recommendation model."""

    id: int
    title: MediaTitle | None = None
    format: MediaFormat | None = None
    type: MediaType | None = None
    status: MediaStatus | None = None
    bannerImage: str | None = None
    coverImage: CoverImage | None = None
    siteUrl: str | None = None


class Recommendation(BaseModel):
    """Recommendation model."""

    id: int
    rating: int | None = None
    userRating: int | str | None = None
    mediaRecommendation: MediaRecommendation | None = None
    user: User | None = None

    @field_validator("userRating", mode="before")
    @classmethod
    def validate_user_rating(cls, v):
        """Convert NO_RATING string to None."""
        if v == "NO_RATING":
            return None
        return v


class RecommendationConnection(BaseModel):
    """Recommendation connection model."""

    edges: list[Any] | None = None
    nodes: list[Recommendation] | None = None
    pageInfo: PageInfo | None = None


# Media relation models
class MediaEdge(BaseModel):
    """Media edge model."""

    id: int | None = None
    relationType: RelationType | None = None
    isMainStudio: bool | None = None
    characters: list[Character] | None = None
    characterRole: CharacterRole | None = None
    characterName: str | None = None
    roleNotes: str | None = None
    dubGroup: str | None = None
    staffRole: str | None = None
    voiceActors: list[VoiceActor] | None = None
    voiceActorRoles: list[Any] | None = None
    node: Media | None = None


class MediaConnection(BaseModel):
    """Media connection model."""

    edges: list[MediaEdge] | None = None
    nodes: list[Media] | None = None
    pageInfo: PageInfo | None = None


# Forward declare Media for use in other models
class Media(BaseModel):
    """Main media model."""

    id: int
    idMal: int | None = None
    title: MediaTitle | None = None
    type: MediaType | None = None
    format: MediaFormat | None = None
    status: MediaStatus | None = None
    description: str | None = None
    startDate: AniListDate | None = None
    endDate: AniListDate | None = None
    season: MediaSeason | None = None
    seasonYear: int | None = None
    seasonInt: int | None = None
    episodes: int | None = None
    duration: int | None = None
    chapters: int | None = None
    volumes: int | None = None
    countryOfOrigin: str | None = None
    isLicensed: bool | None = None
    source: MediaSource | None = None
    hashtag: str | None = None
    trailer: Trailer | None = None
    updatedAt: int | None = None
    coverImage: CoverImage | None = None
    bannerImage: str | None = None
    genres: list[str] | None = None
    synonyms: list[str] | None = None
    averageScore: int | None = None
    meanScore: int | None = None
    popularity: int | None = None
    isLocked: bool | None = None
    trending: int | None = None
    favourites: int | None = None
    tags: list[MediaTag] | None = None
    relations: MediaConnection | None = None
    characters: CharacterConnection | None = None
    characterPreview: CharacterConnection | None = None
    staff: StaffConnection | None = None
    staffPreview: StaffConnection | None = None
    studios: StudioConnection | None = None
    isFavourite: bool | None = None
    isFavouriteBlocked: bool | None = None
    isAdult: bool | None = None
    nextAiringEpisode: NextAiringEpisode | None = None
    airingSchedule: Any | None = None
    trends: Any | None = None
    externalLinks: list[ExternalLink] | None = None
    streamingEpisodes: list[StreamingEpisode] | None = None
    rankings: list[MediaRanking] | None = None
    mediaListEntry: MediaListEntry | None = None
    reviews: ReviewConnection | None = None
    reviewPreview: ReviewConnection | None = None
    recommendations: RecommendationConnection | None = None
    stats: MediaStats | None = None
    siteUrl: str | None = None
    autoCreateForumThread: bool | None = None
    isRecommendationBlocked: bool | None = None
    isReviewBlocked: bool | None = None
    modNotes: str | None = None


# Response models
class MediaResponse(BaseModel):
    """Response model for single media query."""

    media: Media | None = None


class PageResponse(BaseModel):
    """Response model for paginated media query."""

    pageInfo: PageInfo | None = None
    media: list[Media] | None = None


class MediaPageResponse(BaseModel):
    """Response model for media page query."""

    Page: PageResponse | None = None


# Search and query models
class MediaSearchVariables(BaseModel):
    """Variables for media search query."""

    search: str | None = None
    type: MediaType | None = None
    format: list[MediaFormat] | None = None
    status: MediaStatus | None = None
    season: MediaSeason | None = None
    seasonYear: int | None = None
    year: str | None = None
    onList: bool | None = None
    isAdult: bool | None = False
    genre: list[str] | None = None
    tag: list[str] | None = None
    page: int | None = 1
    perPage: int | None = 10


class MediaByIdVariables(BaseModel):
    """Variables for media by ID query."""

    id: int
    type: MediaType | None = None


# Update forward references
Media.model_rebuild()
MediaEdge.model_rebuild()
MediaConnection.model_rebuild()
MediaResponse.model_rebuild()
PageResponse.model_rebuild()
MediaPageResponse.model_rebuild()

File: anyflix-player/src/types/schemas/anilist.ts
/**
 * AniList Zod schemas for GraphQL API responses
 * Complete AniList data models with runtime validation
 */

import { z } from 'zod';
import { 
  MediaFormatSchema, 
  MediaTypeSchema, 
  MediaStatusSchema, 
  MediaSeasonSchema,
  RelationTypeSchema,
  CharacterRoleSchema,
  AniListDateSchema,
  type MediaFormat,
  type MediaType,
  type MediaStatus,
  type MediaSeason,
  type AniListDate
} from './base';

// =============================================
// AniList Core Models
// =============================================

export const MediaTitleSchema = z.object({
  romaji: z.string().optional(),
  english: z.string().optional(),
  native: z.string().optional(),
  userPreferred: z.string().optional()
});
export type MediaTitle = z.infer<typeof MediaTitleSchema>;

export const CoverImageSchema = z.object({
  extraLarge: z.string().optional(),
  large: z.string().optional(),
  medium: z.string().optional(),
  color: z.string().optional()
});
export type CoverImage = z.infer<typeof CoverImageSchema>;

export const TrailerSchema = z.object({
  id: z.string().optional(),
  site: z.string().optional(),
  thumbnail: z.string().optional()
});
export type Trailer = z.infer<typeof TrailerSchema>;

export const NextAiringEpisodeSchema = z.object({
  airingAt: z.number().optional(),
  timeUntilAiring: z.number().optional(),
  episode: z.number().optional()
});
export type NextAiringEpisode = z.infer<typeof NextAiringEpisodeSchema>;

export const MediaTagSchema = z.object({
  id: z.number(),
  name: z.string(),
  description: z.string().optional(),
  category: z.string().optional(),
  rank: z.number().optional(),
  isGeneralSpoiler: z.boolean().optional(),
  isMediaSpoiler: z.boolean().optional(),
  isAdult: z.boolean().optional(),
  userId: z.number().optional()
});
export type MediaTag = z.infer<typeof MediaTagSchema>;

export const MediaRankingSchema = z.object({
  id: z.number(),
  rank: z.number(),
  type: z.string(),
  format: MediaFormatSchema,
  year: z.number().optional(),
  season: MediaSeasonSchema.optional(),
  allTime: z.boolean().optional(),
  context: z.string()
});
export type MediaRanking = z.infer<typeof MediaRankingSchema>;

export const StatusDistributionSchema = z.object({
  status: z.string().optional(),
  amount: z.number().optional()
});
export type StatusDistribution = z.infer<typeof StatusDistributionSchema>;

export const ScoreDistributionSchema = z.object({
  score: z.number().optional(),
  amount: z.number().optional()
});
export type ScoreDistribution = z.infer<typeof ScoreDistributionSchema>;

export const MediaStatsSchema = z.object({
  statusDistribution: z.array(StatusDistributionSchema).optional(),
  scoreDistribution: z.array(ScoreDistributionSchema).optional()
});
export type MediaStats = z.infer<typeof MediaStatsSchema>;

export const StreamingEpisodeSchema = z.object({
  title: z.string().optional(),
  thumbnail: z.string().optional(),
  url: z.string().optional(),
  site: z.string().optional()
});
export type StreamingEpisode = z.infer<typeof StreamingEpisodeSchema>;

export const ExternalLinkSchema = z.object({
  id: z.number(),
  url: z.string().optional(),
  site: z.string(),
  siteId: z.number().optional(),
  type: z.string().optional(),
  language: z.string().optional(),
  color: z.string().optional(),
  icon: z.string().optional(),
  notes: z.string().optional(),
  isDisabled: z.boolean().optional()
});
export type ExternalLink = z.infer<typeof ExternalLinkSchema>;

// =============================================
// Character and Staff Models
// =============================================

export const CharacterNameSchema = z.object({
  first: z.string().optional(),
  middle: z.string().optional(),
  last: z.string().optional(),
  full: z.string().optional(),
  native: z.string().optional(),
  alternative: z.array(z.string()).optional(),
  alternativeSpoiler: z.array(z.string()).optional(),
  userPreferred: z.string().optional()
});
export type CharacterName = z.infer<typeof CharacterNameSchema>;

export const CharacterImageSchema = z.object({
  large: z.string().optional(),
  medium: z.string().optional()
});
export type CharacterImage = z.infer<typeof CharacterImageSchema>;

export const CharacterSchema = z.object({
  id: z.number(),
  name: CharacterNameSchema.optional(),
  image: CharacterImageSchema.optional(),
  description: z.string().optional(),
  gender: z.string().optional(),
  dateOfBirth: AniListDateSchema.optional(),
  age: z.string().optional(),
  bloodType: z.string().optional(),
  isFavourite: z.boolean().optional(),
  isFavouriteBlocked: z.boolean().optional(),
  siteUrl: z.string().optional()
});
export type Character = z.infer<typeof CharacterSchema>;

export const StaffNameSchema = z.object({
  first: z.string().optional(),
  middle: z.string().optional(),
  last: z.string().optional(),
  full: z.string().optional(),
  native: z.string().optional(),
  alternative: z.array(z.string()).optional(),
  userPreferred: z.string().optional()
});
export type StaffName = z.infer<typeof StaffNameSchema>;

export const StaffImageSchema = z.object({
  large: z.string().optional(),
  medium: z.string().optional()
});
export type StaffImage = z.infer<typeof StaffImageSchema>;

export const StaffSchema = z.object({
  id: z.number(),
  name: StaffNameSchema.optional(),
  languageV2: z.string().optional(),
  image: StaffImageSchema.optional(),
  description: z.string().optional(),
  primaryOccupations: z.array(z.string()).optional(),
  gender: z.string().optional(),
  dateOfBirth: AniListDateSchema.optional(),
  dateOfDeath: AniListDateSchema.optional(),
  age: z.number().optional(),
  yearsActive: z.array(z.number()).optional(),
  homeTown: z.string().optional(),
  bloodType: z.string().optional(),
  isFavourite: z.boolean().optional(),
  isFavouriteBlocked: z.boolean().optional(),
  siteUrl: z.string().optional()
});
export type Staff = z.infer<typeof StaffSchema>;

export const VoiceActorSchema = StaffSchema.extend({
  language: z.string().optional()
});
export type VoiceActor = z.infer<typeof VoiceActorSchema>;

// =============================================
// Studio Models
// =============================================

export const StudioSchema = z.object({
  id: z.number(),
  name: z.string(),
  isAnimationStudio: z.boolean().optional(),
  siteUrl: z.string().optional(),
  isFavourite: z.boolean().optional()
});
export type Studio = z.infer<typeof StudioSchema>;

// =============================================
// User Models
// =============================================

export const UserAvatarSchema = z.object({
  large: z.string().optional(),
  medium: z.string().optional()
});
export type UserAvatar = z.infer<typeof UserAvatarSchema>;

export const UserSchema = z.object({
  id: z.number(),
  name: z.string(),
  avatar: UserAvatarSchema.optional(),
  bannerImage: z.string().optional(),
  about: z.string().optional(),
  isFollowing: z.boolean().optional(),
  isFollower: z.boolean().optional(),
  isBlocked: z.boolean().optional(),
  bans: z.array(z.any()).optional(),
  options: z.record(z.any()).optional(),
  mediaListOptions: z.record(z.any()).optional(),
  favourites: z.record(z.any()).optional(),
  statistics: z.record(z.any()).optional(),
  unreadNotificationCount: z.number().optional(),
  siteUrl: z.string().optional(),
  donatorTier: z.number().optional(),
  donatorBadge: z.string().optional(),
  moderatorRoles: z.array(z.string()).optional(),
  createdAt: z.number().optional(),
  updatedAt: z.number().optional()
});
export type User = z.infer<typeof UserSchema>;

// =============================================
// Pagination Models
// =============================================

export const PageInfoSchema = z.object({
  total: z.number().optional(),
  perPage: z.number().optional(),
  currentPage: z.number().optional(),
  lastPage: z.number().optional(),
  hasNextPage: z.boolean().optional()
});
export type PageInfo = z.infer<typeof PageInfoSchema>;

// =============================================
// Connection Types (GraphQL structure)
// =============================================

export const CharacterEdgeSchema = z.object({
  id: z.number().optional(),
  role: CharacterRoleSchema.optional(),
  name: z.string().optional(),
  voiceActors: z.array(VoiceActorSchema).optional(),
  node: CharacterSchema.optional()
});
export type CharacterEdge = z.infer<typeof CharacterEdgeSchema>;

export const StaffEdgeSchema = z.object({
  id: z.number().optional(),
  role: z.string().optional(),
  node: StaffSchema.optional()
});
export type StaffEdge = z.infer<typeof StaffEdgeSchema>;

export const StudioEdgeSchema = z.object({
  id: z.number().optional(),
  isMain: z.boolean(),
  node: StudioSchema.optional()
});
export type StudioEdge = z.infer<typeof StudioEdgeSchema>;

export const CharacterConnectionSchema = z.object({
  edges: z.array(CharacterEdgeSchema).optional(),
  nodes: z.array(CharacterSchema).optional()
});
export type CharacterConnection = z.infer<typeof CharacterConnectionSchema>;

export const StaffConnectionSchema = z.object({
  edges: z.array(StaffEdgeSchema).optional(),
  nodes: z.array(StaffSchema).optional()
});
export type StaffConnection = z.infer<typeof StaffConnectionSchema>;

export const StudioConnectionSchema = z.object({
  edges: z.array(StudioEdgeSchema).optional(),
  nodes: z.array(StudioSchema).optional()
});
export type StudioConnection = z.infer<typeof StudioConnectionSchema>;

// =============================================
// Main Media Model (with lazy evaluation for circular references)
// =============================================

export const AniListMediaSchema = z.lazy(() => z.object({
  id: z.number(),
  idMal: z.number().optional(),
  title: MediaTitleSchema.optional(),
  type: MediaTypeSchema.optional(),
  format: MediaFormatSchema.optional(),
  status: MediaStatusSchema.optional(),
  description: z.string().optional(),
  startDate: AniListDateSchema.optional(),
  endDate: AniListDateSchema.optional(),
  season: MediaSeasonSchema.optional(),
  seasonYear: z.number().optional(),
  seasonInt: z.number().optional(),
  episodes: z.number().optional(),
  duration: z.number().optional(),
  chapters: z.number().optional(),
  volumes: z.number().optional(),
  countryOfOrigin: z.string().optional(),
  isLicensed: z.boolean().optional(),
  source: z.string().optional(),
  hashtag: z.string().optional(),
  trailer: TrailerSchema.optional(),
  updatedAt: z.number().optional(),
  coverImage: CoverImageSchema.optional(),
  bannerImage: z.string().optional(),
  genres: z.array(z.string()).optional(),
  synonyms: z.array(z.string()).optional(),
  averageScore: z.number().optional(),
  meanScore: z.number().optional(),
  popularity: z.number().optional(),
  isLocked: z.boolean().optional(),
  trending: z.number().optional(),
  favourites: z.number().optional(),
  tags: z.array(MediaTagSchema).optional(),
  relations: z.lazy(() => MediaConnectionSchema).optional(),
  characters: CharacterConnectionSchema.optional(),
  characterPreview: CharacterConnectionSchema.optional(),
  staff: StaffConnectionSchema.optional(),
  staffPreview: StaffConnectionSchema.optional(),
  studios: StudioConnectionSchema.optional(),
  isFavourite: z.boolean().optional(),
  isFavouriteBlocked: z.boolean().optional(),
  isAdult: z.boolean().optional(),
  nextAiringEpisode: NextAiringEpisodeSchema.optional(),
  airingSchedule: z.any().optional(),
  trends: z.any().optional(),
  externalLinks: z.array(ExternalLinkSchema).optional(),
  streamingEpisodes: z.array(StreamingEpisodeSchema).optional(),
  rankings: z.array(MediaRankingSchema).optional(),
  mediaListEntry: z.any().optional(),
  reviews: z.any().optional(),
  reviewPreview: z.any().optional(),
  recommendations: z.any().optional(),
  stats: MediaStatsSchema.optional(),
  siteUrl: z.string().optional(),
  autoCreateForumThread: z.boolean().optional(),
  isRecommendationBlocked: z.boolean().optional(),
  isReviewBlocked: z.boolean().optional(),
  modNotes: z.string().optional()
}));

export type AniListMedia = z.infer<typeof AniListMediaSchema>;

export const MediaEdgeSchema = z.object({
  id: z.number().optional(),
  relationType: RelationTypeSchema.optional(),
  isMainStudio: z.boolean().optional(),
  characters: z.array(CharacterSchema).optional(),
  characterRole: CharacterRoleSchema.optional(),
  characterName: z.string().optional(),
  roleNotes: z.string().optional(),
  dubGroup: z.string().optional(),
  staffRole: z.string().optional(),
  voiceActors: z.array(VoiceActorSchema).optional(),
  voiceActorRoles: z.array(z.any()).optional(),
  node: AniListMediaSchema.optional()
});
export type MediaEdge = z.infer<typeof MediaEdgeSchema>;

export const MediaConnectionSchema = z.object({
  edges: z.array(MediaEdgeSchema).optional(),
  nodes: z.array(AniListMediaSchema).optional(),
  pageInfo: PageInfoSchema.optional()
});
export type MediaConnection = z.infer<typeof MediaConnectionSchema>;
